hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/ProtoKafkaSource.java

```diff
@@ -0,0 +1,111 @@
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.streaming.kafka010.KafkaUtils;
+import org.apache.spark.streaming.kafka010.LocationStrategies;
+import org.apache.spark.streaming.kafka010.OffsetRange;
+
+import java.io.Serializable;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+
+/**
+ * Reads protobuf serialized Kafka data, based on a provided class name.
+ */
+public class ProtoKafkaSource extends KafkaSource<Message> {
+
+  private static final Logger LOG = LogManager.getLogger(ProtoKafkaSource.class);
+  // these are native kafka's config. do not change the config names.
+  private static final String NATIVE_KAFKA_KEY_DESERIALIZER_PROP = "key.deserializer";
+  private static final String NATIVE_KAFKA_VALUE_DESERIALIZER_PROP = "value.deserializer";
+  private final String className;
+
+  public ProtoKafkaSource(TypedProperties props, JavaSparkContext sparkContext,
+                          SparkSession sparkSession, SchemaProvider schemaProvider, HoodieDeltaStreamerMetrics metrics) {
+    super(props, sparkContext, sparkSession, schemaProvider, SourceType.PROTO, metrics);
+    DataSourceUtils.checkRequiredProperties(props, Arrays.asList(
+        ProtoClassBasedSchemaProvider.Config.PROTO_SCHEMA_CLASS_NAME));
+    props.put(NATIVE_KAFKA_KEY_DESERIALIZER_PROP, StringDeserializer.class);
+    props.put(NATIVE_KAFKA_VALUE_DESERIALIZER_PROP, ByteArrayDeserializer.class);
+    className = props.getString(ProtoClassBasedSchemaProvider.Config.PROTO_SCHEMA_CLASS_NAME);
+    this.offsetGen = new KafkaOffsetGen(props);
+  }
+
+  @Override
+  JavaRDD<Message> toRDD(OffsetRange[] offsetRanges) {
+    ProtoDeserializer deserializer = new ProtoDeserializer(className);
+    return KafkaUtils.<String, byte[]>createRDD(sparkContext, offsetGen.getKafkaParams(), offsetRanges,
+        LocationStrategies.PreferConsistent()).map(obj -> deserializer.parse(obj.value()));
+  }
+
+  private static class ProtoDeserializer implements Serializable {
+    private final String className;
+    private transient Class protoClass;
+    private transient Method parseMethod;
+
+    public ProtoDeserializer(String className) {
+      this.className = className;
+    }
+
+    public Message parse(byte[] bytes) {
+      try {
+        return (Message) getParseMethod().invoke(getClass(), bytes);
+      } catch (IllegalAccessException | InvocationTargetException ex) {
+        throw new HoodieException("Failed to parse proto message from kafka", ex);
+      }
+    }
+
+    private Class getProtoClass() {
+      if (protoClass == null) {
+        protoClass = ReflectionUtils.getClass(className);
+      }
+      return protoClass;
+    }
+
+    private Method getParseMethod() {
+      if (parseMethod == null) {
+        try {
+          parseMethod = getProtoClass().getMethod("parseFrom", byte[].class);
```

Comment:
should we also handle HoodieException from ReflectionUtils.getClass?

Reply:
Are you suggesting to catch it so we can add some more context to the error? like 
```
catch (HoodieException ex) {
  throw new HoodieException("Unable to get proto class specified: " + className, ex);
}
```

Reply:
I'm just curious whether this needs to be handled, though the stacktrace should be sufficient.